src/cml.js

```diff
@@ -0,0 +1,95 @@
+const GitlabClient = require('./drivers/gitlab');
+const GithubClient = require('./drivers/github');
+const { upload } = require('./utils');
+
+const get_client = (opts) => {
+  const { driver, repo, token } = opts;
+  if (!driver) throw new Error('driver not set');
+
+  if (driver === 'github') return new GithubClient({ repo, token });
+  if (driver === 'gitlab') return new GitlabClient({ repo, token });
+
+  throw new Error('driver unknown!');
+};
+
+class CML {
```

Comment:
Is CML (and `cml.js`) the right name? The code ends up using CML for different objects like:
```
const CML = require('../src/cml');
…
const cml = new CML(opts);
```

One of the big ideas behind this redesign is to separate different concepts - GitHub/GitLab and CI/CDs executors. But it seems like we still using a monolithic CML concept.

I'd suggest renaming this one to something like `ScmHostingDriver` with the factory method and abstract (empty) methods (or without methods since it is JS).

Reply:
If this CML is a general, user-facing API - it makes sense to keep it here and keep the name. But all the client's specific logic needs to be extracted. Including the abstract `*Driver` class with the factory method.